web/text_highlighter.js

```diff
@@ -29,6 +29,10 @@
  * either the text layer or XFA layer depending on the type of document.
  */
 class TextHighlighter {
+  #turnedOff = false;
+
+  #needUpdate = false;
+
```

Comment:
Why do we need all of this additional complexity in this class, and generally having three different but related method just feels very difficult to understand?
Why isn't it sufficient to simply call the existing `enable`/`disable` methods, since it seems like that is actually what you want?

Reply:
The idea is to avoid to uselessly call `this._updateMatches` when it's re-enabled.
About the accessibility manager, the disable method was aimed to be used only when the text layer rendering has been cancelled (and I suppose that it was the same for the highlighter).
Here we want to reuse those functions in an other context: we just want to temporarily disable them, hence it's why I added new ones.
But I can add a parameter to enable/disable to handle that.

Reply:
> The idea is to avoid to uselessly call `this._updateMatches` when it's re-enabled.

It's not really useless though, since the find-state may have changed completely (i.e. having moved to a different page) between the textLayer being hidden/shown depending on the timing of various async operations.

May I suggest the following changes, as far as I understand the situation, since that way we can simply call `enable`/`disable` from the `TextLayerBuilder` and not worry about it?
```diff
diff --git a/web/text_highlighter.js b/web/text_highlighter.js
index 066e00663..c1b828c40 100644
--- a/web/text_highlighter.js
+++ b/web/text_highlighter.js
@@ -95,6 +95,7 @@ class TextHighlighter {
       );
       this._onUpdateTextLayerMatches = null;
     }
+    this._updateMatches(/* reset = */ true);
   }
 
   _convertMatches(matches, matchesLength) {
@@ -264,8 +265,8 @@ class TextHighlighter {
     }
   }
 
-  _updateMatches() {
-    if (!this.enabled) {
+  _updateMatches(reset = false) {
+    if (!this.enabled && !reset) {
       return;
     }
     const { findController, matches, pageIdx } = this;
@@ -283,7 +284,7 @@ class TextHighlighter {
       clearedUntilDivIdx = match.end.divIdx + 1;
     }
 
-    if (!findController?.highlightMatches) {
+    if (!findController?.highlightMatches || reset) {
       return;
     }
     // Convert the matches on the `findController` into the match format
```

Reply:
If I understand correctly, here we're resetting the highlights even if they're all here (I mean there won't be any updates later) and then we're restoring them once the text layer is visible, I still have the impress we're doing something useless.
If I remember correctly, you mentioned that you had a concern with the call to `scrollIntoView` which could give some weird results because the text layer is hidden. So why not just disable the scrollIntoView when the text layer is hidden ?

Reply:
Anyway if you think that your approach is less risky than mine, then I'm fine with it and I'll try to improve things later.
Tell me what you prefer.

Reply:
For me this patch is large and risky enough as-is, see e.g. the search regression found earlier today, and I'd thus very much appreciate if we only make the *minimal changes* required to keep the pre-existing behaviour here :-)